domain/src/com/thoughtworks/go/domain/materials/perforce/P4MaterialUpdater.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2017 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.domain.materials.perforce;
+
+import com.thoughtworks.go.config.materials.perforce.P4Material;
+import com.thoughtworks.go.domain.BuildCommand;
+import com.thoughtworks.go.domain.materials.RevisionContext;
+import com.thoughtworks.go.util.GoConstants;
+import com.thoughtworks.go.util.StringUtil;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.thoughtworks.go.domain.BuildCommand.*;
+import static java.lang.Long.parseLong;
+import static java.lang.String.format;
+
+public class P4MaterialUpdater {
+    private P4Material material;
+    private String clientName;
+
+    public P4MaterialUpdater(P4Material material) {
+        this.material = material;
+    }
+
+    public BuildCommand updateTo(String baseDir, RevisionContext revisionContext) {
+        String workingDir = material.workingdir(new File(baseDir)).getPath();
+        String revision = revisionContext.getLatestRevision().getRevision();
+        this.clientName = material.clientName(new File(workingDir));
+
+        return compose(
+                secret(material.getPassword()),
+                loginIfUsingTickets(),
+                constructClient(workingDir, clientName),
+                cleanWorkingDir(workingDir),
+                echo(format("[%s] Start updating %s at revision %s from %s", GoConstants.PRODUCT_NAME, material.updatingTarget(), revision, material.getServerAndPort())),
+                sync(workingDir, revision),
+                echo(format("[%s] Done.\n", GoConstants.PRODUCT_NAME))
+
+        );
+    }
+
+    private BuildCommand loginIfUsingTickets() {
+       if (material.getUseTickets() && !StringUtil.isBlank(material.getPassword())) {
+           return exec("p4", "login").setExecInput(material.getPassword()).setCommandEnvVars(envVars());
+       } else {
+           return noop();
+       }
+    }
+
+    private Map<String, String> envVars() {
+        Map<String, String> env = new HashMap<>();
+        env.put("P4PORT", material.getServerAndPort());
+        env.put("P4CLIENT", clientName);
+        if (!StringUtil.isBlank(material.getUserName())) {
+            env.put("P4USER", material.getUserName());
+        }
+        if (material.getUseTickets() && !StringUtil.isBlank(material.getPassword())) {
```

Comment:
Comparing [this](https://github.com/gocd/gocd/blob/e116224b74be46f46d7b66950d4b408fcb665df3/domain/src/com/thoughtworks/go/domain/materials/perforce/P4Client.java#L169) and the line here, it looks like this should be:

```
if (!material.getUseTickets() && !StringUtil.isBlank(material.getPassword())) {
```

Without this, `P4PASSWD` is not set.